File: internal/user/registration.go
package user

import (
	"file-fusion-backend/internal/model"
	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/dynamodb"
	"github.com/aws/aws-sdk-go/service/dynamodb/dynamodbattribute"
	"github.com/gin-gonic/gin"
	"golang.org/x/crypto/bcrypt"
	"net/http"
)

func Register(context *gin.Context, session *session.Session) {
	var user model.User
	if err := context.ShouldBindJSON(&user); err != nil {
		context.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	encryptedPassword, _ := bcrypt.GenerateFromPassword([]byte(user.Password), bcrypt.DefaultCost)
	user.Password = string(encryptedPassword)

	db := dynamodb.New(session)

	getInput := &dynamodb.GetItemInput{
		TableName: aws.String("user"),
		Key: map[string]*dynamodb.AttributeValue{
			"username": {
				S: aws.String(user.Username),
			},
		},
	}

	getResult, err := db.GetItem(getInput)
	if err != nil {
		context.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}
	if getResult.Item != nil {
		context.JSON(http.StatusBadRequest, "Username already taken!")
		return
	}

	item, err := dynamodbattribute.MarshalMap(user)
	if err != nil {
		context.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	input := &dynamodb.PutItemInput{
		Item:      item,
		TableName: aws.String("user"),
	}

	if _, err := db.PutItem(input); err != nil {
		context.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	context.JSON(http.StatusOK, user)

}

File: internal/user/registration.py
import hashlib
import json

import boto3 as boto3
from botocore.exceptions import ClientError

dynamodb = boto3.resource('dynamodb')
table = dynamodb.Table('user')


def registration(event, context):
    user = json.loads(event['body'])

    # Hash the password for storage
    user['password'] = hashlib.sha256(user['password'].encode('utf-8')).hexdigest()

    # Check if the username already exists
    try:
        response = table.get_item(Key={'username': user['username']})
        if 'Item' in response:
            return {
                'statusCode': 400,
                'body': json.dumps('Username already taken!')
            }
    except ClientError as e:
        return {
            'statusCode': 500,
            'body': json.dumps(str(e))
        }

    # If username is not taken, store the user
    try:
        table.put_item(Item=user)
        return {
            'statusCode': 200,
            'body': json.dumps(user)
        }
    except ClientError as e:
        return {
            'statusCode': 500,
            'body': json.dumps(str(e))
        }